### PR TITLE
Feedback ctrl

### PR DESCRIPTION
**Pull request by garygra**
Feedback control setup:
  * Adding srv and msgs dependencies
  * Adding sensors to mushr
  * State msg definition
  * Feedback service definition
  * Feedback service sim
  * Reading from the sensors to define the state
  * Adding scripts into Cmake
  * Global frequency variable to handle mujoco/ros timesteps
  * feedback service class
  * simulator uses ros::rate for stepping.
  * Sim viz uses ros::WallRate to always update even if using sim_time
  * keyboard service to *simulate* a feedback controller
  * mushr_feedback implementation
  * launch feedback example

### CODE REVIEW

File: mj_models/CMakeLists.txt [hunk 1]
```diff
@@ -8,7 +8,7 @@ set(CMAKE_CXX_STANDARD 17)
 
 find_package(catkin REQUIRED COMPONENTS
   roscpp
-  std_msgs
+  std_msgs geometry_msgs
```
Comment (aravindsiv) [inline mj_models/CMakeLists.txt]: Nit: New line for readability

File: mj_models/CMakeLists.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.0.2)
+cmake_minimum_required(VERSION 3.5)
```
Comment (aravindsiv) [inline mj_models/CMakeLists.txt:1]: Is there any specific reason we need `3.5` instead of maybe `3.2`? I

File: mj_models/msg/mushr/MushrState.msg [hunk 3]
```diff
@@ -0,0 +1,4 @@
+std_msgs/Header header
+geometry_msgs/Pose pose
+std_msgs/String[] str_extra # For additional non-pose data
```
Comment (aravindsiv) [inline mj_models/msg/mushr/MushrState.msg]: Why do we have this?

File: mj_models/msg/mushr/MushrState.msg
Comment (aravindsiv) [inline mj_models/msg/mushr/MushrState.msg:1]: It would be more appropriate to name this something like `MushrObservation` or `MushrObservedState` or `MushrEstimatedState`.

File: mujoco_ros/include/mujoco_ros/defs.h [hunk 4]
```diff
@@ -6,4 +8,13 @@
...
-#include "ros/ros.h"
\ No newline at end of file
+#include "ros/ros.h"
+#include "mujoco/mujoco.h"
+
+namespace mj_ros
+{
+namespace constants
+{
+static double frequency{ 50.0 };
```
Comment (aravindsiv) [inline mujoco_ros/include/mujoco_ros/defs.h]: As we discussed offline, maybe this should not be a constant and set instead as `1.0/m->opt.timestep`.

File: mujoco_ros/include/mujoco_ros/simulator.hpp [hunk 5]
```diff
@@ -18,6 +17,8 @@ class mujoco_simulator_t : public std::enable_shared_from_this<mujoco_simulator_
 
   GLFWwindow* window;
 
+  double _frequency;
```
Comment (aravindsiv) [inline mujoco_ros/include/mujoco_ros/simulator.hpp]: We can remove this, right?

File: mj_models/include/mj_models/mj_mushr.hpp [hunk 6]
```diff
@@ -1,15 +1,28 @@
...
+#include <mj_models/MushrState.h>
+#include <mj_models/MushrFeedback.h>
+
 namespace mj_models
 {
 
 struct mushr_t
 {
   static constexpr std::size_t u_dim{ 2 };
+  struct sensors_t
```
Comment (aravindsiv) [inline mj_models/include/mj_models/mj_mushr.hpp:16]: Maybe something more descriptive, like `ground_truth_pose_sensor`.

File: mj_models/include/mj_models/mj_mushr.hpp [hunk 7]
```diff
@@ -20,4 +33,16 @@ inline void copy(Ctrl ctrl_out, const mj_models::MushrControl& msg)
   ctrl_out[1] = msg.velocity.data;
 }
 
+template <typename Sensor>
+inline void state_from_sensors(mj_models::MushrState& msg, const Sensor& sensordata)
```
Comment (aravindsiv) [inline mj_models/include/mj_models/mj_mushr.hpp]: See above -- since these will not be the sensors that we will be using for the real system (or even a simulated version in experiments), we should ideally make the distinction between the ground truth pose and what a sensor provides.

**Review by aravindsiv — CHANGES_REQUESTED** [review]
I will check the feedback files logic, but in the meantime, there are some minor changes to address.

File: mj_models/CMakeLists.txt (hunk 2 above)
Comment (garygra) [reply mj_models/CMakeLists.txt:1]: It might be only my machine, but I get:
```
CMake Deprecation Warning at mujoco_ros/mujoco_ros/CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.
```

File: mj_models/msg/mushr/MushrState.msg
Comment (garygra) [reply mj_models/msg/mushr/MushrState.msg:1]: I like MushrObservation

File: mj_models/include/mj_models/mj_mushr.hpp (hunk 6 above)
Comment (garygra) [reply mj_models/include/mj_models/mj_mushr.hpp:16]: I intended this to be for *any* sensor, not only pose (i.e. IMU)

File: mj_models/msg/mushr/MushrObservation.msg [hunk 8]
```diff
@@ -0,0 +1,4 @@
+std_msgs/Header header
+geometry_msgs/Pose pose
+std_msgs/String[] str_extra # For additional non-pose data
```
Comment (aravindsiv) [inline mj_models/msg/mushr/MushrObservation.msg:3]: What are these for?

File: mujoco_ros/scripts/keyboard_service.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+screen = None
+control = MushrControl()
+observation = MushrObservation()
+robot_name = "mushr"
+velocity_increase = 0.1
+steering_increase = 0.1
+
+def handle_service(req):
+    global control
```
Comment (aravindsiv) [inline mujoco_ros/scripts/keyboard_service.py]: I am not a fan of Python `global`. It makes sense to put all of these in a Python class. Let me work on that.

File: mujoco_ros/launch/mushr_feedback_example.launch [hunk 10]
```diff
@@ -0,0 +1,10 @@
+<launch>
+    <param name="model_path" type="string" value="$(find mj_models)/models/mushr/mushr.xml"/>
+    <group ns="mushr">
+        <node pkg="mujoco_ros" name="mushr_feedback" type="mushr_feedback" output="screen" >
+            <param name="visualize" value="true" />
+        </node>
+        <!-- <node pkg="mujoco_ros" name="keyboard_service" type="keyboard_service.py" output="screen" launch-prefix="xterm -e"/> -->
```
Comment (aravindsiv) [inline mujoco_ros/launch/mushr_feedback_example.launch]: Why is this removed?

**Review by aravindsiv — CHANGES_REQUESTED** [review]
Only some minor comments that we can discuss offline, and then I will make a change to the keyboard service.

File: mujoco_ros/launch/mushr_feedback_example.launch (hunk 10 above)
Comment (garygra) [reply mujoco_ros/launch/mushr_feedback_example.launch]: In some setups, is easier to launch this in a separate terminal (and the launch-prefix for a separate terminal not always works)

**Review by aravindsiv — APPROVED** [review]
LGTM, we can merge.
